### PR TITLE
More SmartPointer tests

### PR DESCRIPTION
**Pull request by ckerr**
* Test the sample code for memory leaks/errors
 * Add a default constructor for SmartPointer, which adds more edge cases the coder must handle
 * _(minor)_ Fix return value of SmartPointer operator=()
 * _(minor)_ Silence compiler warning in macros.h

### CODE REVIEW

File: lessons/smart-pointers/Makefile [hunk 1]
```diff
@@ -18,7 +18,8 @@ build/main.o: src/main.cpp
 	${COMPILE} -o build/main.o src/main.cpp
 
 test: build/main
-	diff <(./build/main) out.txt && echo SUCCESS || echo FAIL
+	diff <(./build/main) out.txt && echo "PASS OUTPUT" || echo "FAIL OUTPUT"
+	valgrind --error-exitcode=1 --quiet --leak-resolution=low --leak-check=full --show-possibly-lost=no ./build/main > /dev/null && echo "PASS VALGRIND" || echo "FAIL VALGRIND"
```
Comment (groundwater) [inline lessons/smart-pointers/Makefile]: I'm hesitant to introduce too many new things at once. If we use valdrind with electron I'd be happy to have another lesson. Do we use it?

File: lessons/smart-pointers/src/main.cpp [hunk 2]
```diff
@@ -26,6 +26,7 @@ class SmartPointer {
 private:
     PointerBox* pb;
 public:
+    SmartPointer(): pb(nullptr) {}
```
Comment (groundwater) [inline lessons/smart-pointers/src/main.cpp]: Let's leave the implementation out. That's part of the exercise.

File: lessons/smart-pointers/src/main.cpp [hunk 3]
```diff
@@ -85,5 +86,26 @@ int main() {
         SmartPointer spA = make(7);
     }
 
+    TEST(6)
+    {
+        SmartPointer spA;
```
Comment (groundwater) [inline lessons/smart-pointers/src/main.cpp]: Tricky! I like this.

File: lessons/smart-pointers/Makefile (hunk 1 above)
Comment (ckerr) [reply lessons/smart-pointers/Makefile]: Unfortunately not because of issues between Chromium and valgrind described at https://github.com/electron/electron/issues/11769.... however I see there's still [documentation](https://www.chromium.org/nativeclient/how-tos/debugging-documentation/running-under-valgrind) at chromium.org on how to use valgrind, so maybe this is worth looking into again.

Anyway as to this PR, IMO Electron's use of valgrind is beside the point because the person working through these exercises only has to run "make test" and doesn't need to understand anything about valgrind. And since the exercise is about heap memory management, this is an important piece of the puzzle.

File: lessons/smart-pointers/src/main.cpp (hunk 2 above)
Comment (ckerr) [reply lessons/smart-pointers/src/main.cpp]: :+1:

File: lessons/smart-pointers/Makefile (hunk 1 above)
Comment (ckerr) [reply lessons/smart-pointers/Makefile]: _(edit)_ though, looking at this again I did mix two unrelated topics into the same PR. If you like the new tests in main.cpp but are less sure about valgrind, I can split that out so that you don't have to take them as a set
